Improve performance when calculating chat context initial load

fluiddot opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Related to https://github.com/Automattic/studio/pull/252#pullrequestreview-2127960814 and https://github.com/Automattic/dotcom-forge/issues/7835.

## Proposed Changes

- Mark the initial load of chat context complete before fetching data. This will ensure we only do the initial load once per site.
- Create separate process for running inline `wp-cli` commands. The logic and structure is similar to what we did for running sites (https://github.com/Automattic/studio/pull/19).
- Update path functions to support child processes. This was originally changed in https://github.com/Automattic/studio/pull/19/commits/d37aecccadfaa4055474ce3252f1a1774d5d9428 but now due to having more child processes, we need a bit more generic solution.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

### Performance

**Preparation:**
- Create 5 or more sites.

**Steps:**
- Run the app with the command `STUDIO_AI=true npm start`.
- Starting from the first site, select it and continue selecting the rest of the sites quickly.
- Observe the performance of the app is not degraded.

### Site context in AI Assistant

- Run the app with the command `STUDIO_AI=true npm start`.
- Select any site.
- Navigate to the AI Assistant tab.
- Send the prompt `Give me the site context`.
- Observe that the AI responds with expected values associated with the selected site.
- Select another site.
- Add and activate a plugin (e.g. Classic Editor).
- Add and activate a theme (e.g. Seedlet).
- Go back to the app.
- Send the prompt `Give me the site context`.
- Observe that the AI responds with expected values associated with the selected site.

### Server site

- Run the app with the command `STUDIO_AI=true npm start`.
- Start a site.
- Navigate to the site and WP-Admin
- Observe the site works as expected.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Have you checked for TypeScript, React or other console errors?

fluiddot committed:
Mark chat context load complete right after `run` is triggered

In case any of the `wp-cli` commands fail, we'll set the initial load as not complete.

fluiddot committed:
Update path functions to support child processes

fluiddot committed:
Create child process for `wp-cli` commands

fluiddot committed:
Use separate process in `executeWPCLiInline` IPC handler

fluiddot committed:
Only log new session message in main process

fluiddot committed:
Remove unneeded imports from `executeWPCli`

fluiddot commented on src/lib/wp-cli-process-child.ts:
In the future, it would be great to modularize the logic for creating and setting up a separate process. For now, we'll have some redundancy between this process and the site server.

fluiddot commented on src/lib/wp-cli-process.ts:
```diff
@@ -0,0 +1,139 @@
+		} );
+	}
+
+	async stop() {
+		await this.#killProcess();
+	}
```
I left this function in case we want to stop the process before it finishes. E.g. when the execution should be superseded by another one.

sejas reviewed (approved):
I tested it and I don't see the performance issue anymore while changing the sites. The site context is still working.

Great job! Thanks for improving the performance so fast.


https://github.com/Automattic/studio/assets/779993/429805fe-785b-45c7-bea9-82d7bad921c8

fluiddot committed:
Merge branch 'trunk' into fix/chat-context-calculation-performance

# Conflicts:
#	src/storage/paths.ts

fluiddot committed:
Merge branch 'trunk' into fix/chat-context-calculation-performance

fluiddot commented:
The CI job `buildkite/studio/e2e-tests-on-mac` is failing consistently (I retried it three times so far). I'll take a look before merging the PR in case it's uncovering an issue.

dcalhoun commented:
> The CI job `buildkite/studio/e2e-tests-on-mac` is failing consistently (I retried it three times so far). I'll take a look before merging the PR in case it's uncovering an issue.

FWIW, I experienced seemingly [flaky failures](https://buildkite.com/automattic/studio/builds/1407#019035d6-8fd8-458c-b20d-9208708965c2/653-689) in https://github.com/Automattic/studio/pull/274. In my case, it passed on my first retry.

fluiddot committed:
Fix issue after solving conflicts with trunk

The app data path provided by the child process should precede over the ones provided by E2E tests.

fluiddot commented:
> FWIW, I experienced seemingly [flaky failures](https://buildkite.com/automattic/studio/builds/1407#019035d6-8fd8-458c-b20d-9208708965c2/653-689) in #274. In my case, it passed on my first retry.

I think in this case it's an actual failure caused by a mistake when [solving conflicts](https://github.com/Automattic/studio/pull/272/commits/d686ce5f56db8b540ff0e6606642134688da8981). I've addressed it in https://github.com/Automattic/studio/pull/272/commits/3af018d485aeba7f78adba8108406701ea240e1e.
